Add functions to AdjRIBIn/AdjRIBOut for RIS

sebageek opened the pull request:
These functions are required so AdjRIBIn and AdjRIBOut can be used the
same way as LocRIB. With this we can treat the RIB of a BMP server the
same way as the RIB of a BGP server.

---
To use the RIB of BMP and BGP in RIS I need to extend this interface. There is also the option of putting the Interface into the RIS or adding another interface to the routingtable package. In any case we would need to add `Get()`, `LPM()` and `GetLonger()` to be exposed.

I'm very open for input on this one.

BarbarossaTM commented on protocols/bgp/server/update_sender.go:
```diff
@@ -383,3 +383,18 @@ func (a *UpdateSender) ReplacePath(*net.Prefix, *route.Path, *route.Path) {
+
+// LPM is here to fulfill an interface
+func (a *UpdateSender) LPM(pfx *net.Prefix) (res []*route.Route) {
+	return nil
```
To be consistent with above there should be a warning like

`log.Warningf("BGP Update Sender: RouteCount not implemented")`

sebageek replied:
Done

BarbarossaTM commented on protocols/kernel/kernel.go:
```diff
@@ -78,3 +78,18 @@ func (k *Kernel) ReplacePath(*net.Prefix, *route.Path, *route.Path) {
+
+// Get is here to fulfill an interface
+func (k *Kernel) Get(pfx *net.Prefix) *route.Route {
+	return nil
```
This seems to be ok for now, but I guess the kernel RT needs some love.

hikhvar commented on routingtable/client_interface.go:
```diff
@@ -13,10 +13,14 @@ type RouteTableClient interface {
 	Dump() []*route.Route
 	ReplaceFilterChain(filter.Chain)
 	RefreshRoute(*net.Prefix, []*route.Path)
+	Get(pfx *net.Prefix) *route.Route
```
Why do we need to increase the surface of this interface? Is the interface used somewhere in the RIS? 
I prefer smaller interfaces. If we want to define a general interface to plugin the AdjRIBIn or AdjRIBOut into the RIS, we should define a smaller interface in the RIS package. In Go we should define the smallest interface as possible in the consumer package.

sebageek replied:
I'm currently implementing a version of the ris that uses both BMP and BGP as a data source. For this I need to have an interface that covers LocRIB and AdjRIBIn, but this is not yet used anywhere in current master. Extending the RouteTableInterface was one possible option to solve my problem.

I talked to @taktv6 off-github about initially and he said he would be fine with this proposed solution, as he wants to refactor the RIB* implementations anyway. If you on the other hand do not feel comfortable with this I can change this PR to only add `RegisterWithOptions()`, `Get()`, `GetLonger()` and `LPM()` to AdjRIBIn (and for the sake of consistency to AdjRIBOut), leafe the RouteTableClient interface alone and create my own little interface with just these four functions I actualy need in the ris package.

hikhvar replied:
Since it is no problem in golang to implement multiple interfaces I prefer a new smaller interface in the RIS package. 
> and for the sake of consistency to AdjRIBOut
It may be consistent but does it have any meaning? If AdjRIBOut is not a meaningful RIS table client, why support it?

sebageek replied:
I don't need it ith AdjRIBOut, but it could be used there, as it is a RoutingTable member. For example one could ask the AdjRIBOut what routing would look like for the RIB advertised towards another peer. The "sake of consistency" was ment as "if I'm already adding those methods to AdjRINBIn (which I need for the RIS-case) I could also add them to AdjRIBOut (where I don't need them, but they could be useful to somebody else). Does that make sense? If you'd prefer to keep the PR minimal here as well I can also only explicitly add the function I require for my RIS-use-case.

hikhvar replied:
Sorry, had the wrong struct in mind. The AdjRIBOut does make some sense. I thought of the update sender :facepalm:

sebageek replied:
As title, commit message and code is a bit different I have opened a new PR as an alternative option. Does #268 look better? (If you prefer I can also push the changes into this PR)

hikhvar replied:
I like the other MR! It keeps the RouteTableClient interface smaller. If there is a need to add more functions to the interface we can do it later.

sebageek replied:
Cool :) Should I push the changes into this PR?

sebageek committed:
Add functions to AdjRIBIn/AdjRIBOut for RIS

The funtions RegisterWithOptions(), LPPM(), Get() and GetLonger() need
to be exposed by AdjRIBIn/AdjRIBOut. This makes these RIBs usable in
much the same way as LocRIB for the RIS.

sebageek commented:
Moved changes from #268 to here and closed the alternative. As mentioned in the other PR, the interface for the could look something like this then (but I'd keep that in the RIS part of the code)
```golang
type RISRIB interface {
    RegisterWithOptions(routingtable.RouteTableClient, routingtable.ClientOptions)
    Unregister(routingtable.RouteTableClient)
    Dump() []*route.Route
    LPM(pfx *net.Prefix) (res []*route.Route)
    Get(pfx *net.Prefix) *route.Route
    GetLonger(pfx *net.Prefix) (res []*route.Route)
}
```
